src/plugins/visualizations/public/embeddable/visualize_embeddable.ts

```diff
@@ -465,4 +505,87 @@ export class VisualizeEmbeddable
+   */
+  public async populateVisLayers(): Promise<void> {
+    const visLayers = await this.fetchVisLayers();
+    this.visLayers =
+      this.visAugmenterConfig?.visLayerResourceIds === undefined
+        ? visLayers
+        : visLayers.filter((visLayer) =>
+            this.visAugmenterConfig?.visLayerResourceIds?.includes(visLayer.pluginResource.id)
+          );
+  }
+
+  /**
+   * Collects any VisLayers from plugin expressions functions
+   * by fetching all AugmentVisSavedObjects that match the vis
+   * saved object ID.
+   */
+  fetchVisLayers = async (): Promise<VisLayers> => {
+    try {
+      const expressionParams: IExpressionLoaderParams = {
+        searchContext: {
+          timeRange: this.timeRange,
+          query: this.input.query,
+          filters: this.input.filters,
+        },
+        uiState: this.vis.uiState,
+        inspectorAdapters: this.inspectorAdapters,
+      };
+      const aborted = get(this.abortController, 'signal.aborted', false) as boolean;
+      const augmentVisSavedObjs = await getAugmentVisSavedObjs(
+        this.vis.id,
+        this.savedAugmentVisLoader
+      );
+
+      if (!isEmpty(augmentVisSavedObjs) && !aborted && isEligibleForVisLayers(this.vis)) {
+        const visLayersPipeline = buildPipelineFromAugmentVisSavedObjs(augmentVisSavedObjs);
+        // The initial input for the pipeline will just be an empty arr of VisLayers. As plugin
+        // expression functions are ran, they will incrementally append their generated VisLayers to it.
+        const visLayersPipelineInput = {
+          type: 'vis_layers',
+          layers: [] as VisLayers,
+        };
+        // We cannot use this.handler in this case, since it does not support the run() cmd
+        // we need here. So, we consume the expressions service to run this directly instead.
+        const exprVisLayers = (await getExpressions().run(
+          visLayersPipeline,
+          visLayersPipelineInput,
+          expressionParams as Record<string, unknown>
+        )) as ExprVisLayers;
+        const visLayers = exprVisLayers.layers;
+
+        // There may be some stale saved objs if any plugin resources have been deleted since last time
+        // data was fetched from them via the expression functions. Execute this trigger so any listening
+        // action can perform cleanup.
+        getUiActions().getTrigger(PLUGIN_RESOURCE_DELETE_TRIGGER).exec({
+          savedObjs: augmentVisSavedObjs,
+          visLayers,
+        });
+
+        const err = getAnyErrors(visLayers, this.vis.title);
+        // This is only true when one or more VisLayers has an error
+        if (err !== undefined) {
+          const { toasts } = getNotifications();
+          toasts.addError(err, {
+            title: i18n.translate('visualizations.renderVisTitle', {
+              defaultMessage: `Error loading data on the ${this.vis.title} chart`,
+            }),
+            toastMessage: ' ',
+            id: this.id,
```

Comment:
can we add `id` as an optional param to `ErrorToastOptions`

Reply:
It is a separate effort that is already in `2.x`, it needs to be front-ported to `main`. cc owner @ashwin-pc

Reply:
Original discussion: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4414#discussion_r1244423148